src/basicfuns.jl

```diff
@@ -190,12 +190,12 @@ function logsumexp(X)
     isempty(X) && return log(sum(X))
     reduce(logaddexp, X)
 end
-function logsumexp(X::AbstractArray{T}) where {T<:Real}
+function logsumexp(X::AbstractArray{T}; dims=:) where {T<:Real}
     isempty(X) && return log(zero(T))
-    u = maximum(X)
-    isfinite(u) || return float(u)
+    u = maximum(X; dims=dims)
+    u isa AbstractArray || isfinite(u) || return float(u)
     let u=u # avoid https://github.com/JuliaLang/julia/issues/15276
-        u + log(sum(x -> exp(x-u), X))
+        u .+ log.(sum(exp.(X .- u); dims=dims))
```

Comment:
This is going to allocate a temporary array. Unfortunately, there's no way currently to compute the sum over dimensions of a lazy `Broadcasted` object: we need https://github.com/JuliaLang/julia/pull/31020 for this. Until then, better keep a branch with the existing code when `u` is an array, and leave a TODO.

Reply:
By "a branch" you mean an `if` statement or something?

Reply:
This point is still relevant: please keep the current code when `u` isn't an array, as it's more efficient. Also leave a TODO.